### ISSUE TITLE
Memory leak in `Queue##processJobs`

### ISSUE THREAD

**Issue by jbourassa** [OptimalBits/bull#243]
Hey,

I have been investigating a memory leak in our workers and I think it's coming from Bull. I came down to a very minimal scenario that leaks, even without adding any job to the queue:

```javascript 
var queue = Queue('some-queue');
queue.process(200, function(job, done) {
  done();
});
```

If you leave this code running long enough, you'll end up with leaking promises. The fact that leaked strings are almost exclusively `bull:scrape-job:wait` and `bull:scrape-job:active` makes me believe that the leak is coming from `Queue##processJobs` — here's a heap dump comparison:

![screen shot 2016-02-09 at 12 10 34 pm](https://cloud.githubusercontent.com/assets/224311/12924073/5d50b286-cf26-11e5-90c7-8bf3d8b1a689.png)

I've dug through the Queue code and I don't really see why this is leaking. Am I missing something?

**Comment by sam1360** [comment]
Are you ever dumping the Queues out of redis using clean? That could be your issue since the jobs will stay in redis till you call clean.

**Comment by jbourassa** [comment]
@sam1360 I am, but this is unrelated; the screenshot is a heap dump of node-inspector showing that memory usage keeps increasing over time, and that there's a bunch of `bull:*` strings that are leaking.

**Comment by manast** [comment]
@jbourassa not sure the strings are leaks. I think V8 caches the strings even after garbage collecting... I tried to find info about it but couldnt find it, but I have similar stuff when debugging memory leak elsewhere.

**Comment by jbourassa** [comment]
@manast Interesting, maybe I'm following a wrong lead. I'm assuming you are not seeing leaks in your bull usage.

I ran the code within the context of my app. I'll try again with a blank app/context, let it run for a while and take a few snapshots. I guess if the memory keeps going up it'll be a sign that the leak is truly in bull. I'll report back.

**Comment by manast** [comment]
@jbourassa do you the three snapshot method for finding leaks? https://www.inqbation.com/website-performance-enhancements-fix-memory-leaks/

I use it when I need to find leaks, I have not investigate them on bull for some time, so it is fully possible some leak has been introduced lately. We should have some regression test for leaks as well... but it is quite time consuming.

**Comment by jbourassa** [comment]
@manast Yes, I'm using multiple snapshots. I dug a little more this morning and could indeed reproduce the promise leak with only bull running on idle (without any jobs being posted).

As far as I know, this code change does not impact the code behaviour in any way — the special behaviour of the `finally` method isn't used — but stops promises from leaking:

```diff
diff --git a/lib/queue.js b/lib/queue.js
index 42d07e1..149c379 100644
--- a/lib/queue.js
+++ b/lib/queue.js
@@ -528,6 +528,7 @@ Queue.prototype.processStalledJob = function(job){
 
 Queue.prototype.processJobs = function(){
   var _this = this;
+  var processJobs = this.processJobs.bind(this);
 
   return this.processStalledJobs().then(function() {
     return _this.getNextJob();
@@ -537,9 +538,7 @@ Queue.prototype.processJobs = function(){
         return _this.processJob(job);
       }
     });
-  }).finally(function(){
-    return _this.processJobs();
-  });
+  }).then(processJobs, processJobs);
 };
 
 Queue.prototype.processJob = function(job){
```

I'll investigate further but looks like the issue might be in Bluebird. Feel free to comment if you think I'm wrong in any way.

**Comment by jozsi** [comment]
@jbourassa - if you suspect the Promise library (`bluebird`) in this case, you cloud give [any-promise](https://github.com/kevinbeaty/any-promise) a spin to easily switch between Promise implementations.

**Comment by jbourassa** [comment]
Closed in favour of #245.
